### ISSUE TITLE
Data spilling in join

### ISSUE THREAD

**Issue by kokosing** [prestodb/presto#5897]
We started to work on data spilling in join. Currently it was mostly conceptual work, not much code yet.

Before we start any serious coding we would like to hear you opinion about our idea of how we think it should be done.

Here is a design doc: https://docs.google.com/document/d/1FyT2-2ocOZ6dvzGK70yrUIhd9oucmfnreFWqd3SmfyU/edit#

If you notice any problem: our code misunderstanding, performance or correctness issue, please comment there.

Our idea is to arrange short meeting once you review above document to discuss that design. Unless everything is obvious and clear then we will just start to code ;)

CC: @pnowojski @losipiuk @martint, @dain, @erichwang, @cberner  (feel free to invite others if I missed someone with context)

**Comment by cberner** [comment]
Overall that looks like a reasonable approach. I do have one concern though. Both this and the spilling for aggregations look like they're going to add a lot of complexity to the query execution. 

Something I've been considering for a while is "contingency planning" where the planner inserts parts of the plan which are only taken under certain conditions, and I wonder if it would work well here.

I would envision it working something roughly like this:
Instead of 
```
          Join
         /      \
       TS     HashBuild
                   |
                  TS
```

You would have something like:
```
                                    Join
                                  /        \
     Join                HashBuild          |
    /     \                    |            |
  TS    HashBuild     PartitionToDisk PartitionToDisk  
                 \               /              |
                 SpoolAndMeasureSize           TS
                          |
                         TS
```
* ignore the fact that the hash build goes into different sides of the join in the two cases. I just couldn't make the ascii art work out.

The idea would be that you scan the build side, buffer it until you know whether it's going to fit in memory or not and then execute one join strategy or the other. I think this might lead to a cleaner implementation, as it would separate the responsibility of operators better. Another benefit is that down the road it may be possible to have information communicated back to the coordinator, and have the coordinator trigger contingencies. For example, dynamically adjusting parallelism to speed-up bottlenecks in the plan.

Let me know what you think. I haven't thought this through in a lot of detail, but it's an issue we keep running into for various execution plans.

**Comment by kokosing** [comment]
@cberner Great idea! I think it should work just well for join. It already had something like that somewhere in `SpillableLookupSourceSupplier`, it had to take the decision whether to spill or not and then run different path of code for each case. However it was not as clearly designed as with your approach will be.

Although, I don't think it fit so nicely to spilling for aggregation. Aggregation can processes very large tables without a need for spilling to disk. Decision about doing the spill is not based on size of read pages, but on number of distinct aggregation groups. To calculate this number you need to do the aggregation. To make use of your idea we would need to extract some aggregation work to SpoolAndMeasureSize (maybe as kind of a compaction function) and then do another part of aggregation work in subsequent spilling or in memory aggregation operators. It sounds more difficult than for join, but it looks doable. But the question is: is that abstraction better than having that logic enclosed in AggregationOperator internals?

The other problem is that: how this relates to revocable memory (https://github.com/prestodb/presto/pull/5843)? We envisioned that at some point most of the operators would use this kind of memory for joins, aggregation and others. In that way, in case of long running (low priority) query `S`, when another query `F` comes in we could reclaim the memory from `S` to `F` by revoking it. To achieve that `S` would need to spill the intermediate data on disk and thanks to that `F` could finish faster (with lower likelihood of spill). With your approach that would mean that `SpoolAndMeasureSize` should spill data to disk and so it would need to know how to partition the data (in can be injected).

Let me think it about it more. I would like to incorporate your approach into document and implement that for spilling data for join, but in order to do so we need fit this with revocable memory and aggregation in clear fashion.

**Comment by cberner** [comment]
Yep, I think you're right that this won't work well for aggregations.

I haven't thought about how this would fit into revocable memory. One issue I see is that it's not clear what action should be taken by an operator when its memory is revoked. For example, in the case of an aggregation becoming too large to fit in memory it clearly needs to switch to merge-sort. However, in the case of a higher priority query preempting an aggregation, it might be better to spill to disk and wait for the memory to become free then page back in and continue.

**Comment by pnowojski** [comment]
@cberner I don't see how this could work with revocable memory, because revoking can be triggered from the outside, even based on conditions completely unrelated to the spilling query. So in the end we still need some kind of spilling mechanism in both `HashBuild` and `HashJoin` operators.

>  For example, in the case of an aggregation becoming too large to fit in memory it clearly needs to switch to merge-sort. However, in the case of a higher priority query preempting an aggregation, it might be better to spill to disk and wait for the memory to become free then page back in and continue.

In aggregation you don't have to such things. If query A is spilled because of higher priority query B, when query B finishes query A doesn't need to page back. It just can continue working in memory. Once finished, it just need to merge sort previously spilled content with in memory content.

**Comment by fmeiser** [comment]
I think it would be easier to improve the current implementation of partitioned hash join from @dain. If it creates many independent small partitions Ri and join Ri only with the corresponding partition Si then the memory footprint would relativ small and you improve cache hits in probe phase (see Shatdal et al). One challenge will be a smart buffering on probe side to avoid small pages for Si.
